---
title: "Class 9: Structural Bioinformatics (Pt. 1)"
author: "Nataliana Hernandez (PID: A17096549)"
format: html
---

The main database for structural data is called the PDB (Protein Data Bank). Let's see what it contains: 

Data from : http://www.rcsb.org/

Read this into R:
```{r}
pdbdb <- read.csv("Data Export Summary.csv")
pdbdb
```
and answer the following questions:


```{r}
pdbdb$Total
```

I need to remove the comma and convert to numeric to do math: 
```{r}
as.numeric( sub(",","", pdbdb$Total) )
```

I could turn this into a function to fix the whole table or any future table I read like this: 

```{r}
x <- pdbdb$Total
as.numeric( sub(",","", x) )
```

```{r}
comma2numeric <- function(x) {
  as.numeric( sub(",","", x) )
  }
```

Test it
```{r}
comma2numeric(pdbdb$X.ray)
```

```{r}
apply(pdbdb, 2, comma2numeric)
```


## Or try a different read/import function:

```{r}
#/ message: false
library(readr)
pdbdb <- read_csv("Data Export Summary.csv")
```

```{r}
sum(pdbdb$Total)
```

>Q1: What percentage of structures in the PDB are solved by X-Ray and Electron Microscopy.

```{r}
sum(pdbdb$`X-ray`)/sum(pdbdb$Total) * 100
```
```{r}
sum(pdbdb$EM)/sum(pdbdb$Total) * 100
```


>Q2: What proportion of structures in the PDB are protein?

```{r}
pdbdb$Total[1]/sum(pdbdb$Total) * 100
```



## Mol*

Mol* (pronounced "molstar") is a new web-based molecular viewer that we will need to learn the basics of here.

https://molstar.org/viewer/

We will use PDB code: 1HSG

![A first image from molstar](1HSG.png)

Some more custom images:

![The all important catalytic ASP25 amino acids](1HSG.2.png)


![](1HSG.3.png)


## The Bio3D package


```{r}
#readLines("hsg.pdb")
```



>Q3: Type HIV in the PDB website search box on the home page and determine how many HIV-1 protease structures are in the current PDB?

- 5 structures


>Q4: Water molecules normally have 3 atoms. Why do we see just one atom per water molecule in this structure?

- The water molecules are only represented by the oxygen atom in the entire molecule, rather than all three atoms present based on x-ray crystallography. 

>Q5: There is a critical “conserved” water molecule in the binding site. Can you identify this water molecule? What residue number does this water molecule have

- The conserved water molecule is the one in the binding site that stabilizes the interaction. HOH 308. 


>Q6: Generate and save a figure clearly showing the two distinct chains of HIV-protease along with the ligand. You might also consider showing the catalytic residues ASP 25 in each chain and the critical water (we recommend “Ball & Stick” for these side-chains). Add this figure to your Quarto document.

![Both ASP 25](1HSG.4.png)

>Discussion Topic: Can you think of a way in which indinavir, or even larger ligands and substrates, could enter the binding site?


```{r}
library(bio3d)
```

```{r}
pdb <- read.pdb("1hsg")
pdb
```


```{r}
attributes(pdb)
```

```{r}
head(pdb$atom)
```


>Q7: How many amino acid residues are there in this pdb object? 

- 198

>Q8: Name one of the two non-protein residues? 

- HOH (127) and MK1 (1)

>Q9: How many protein chains are in this structure?

- 2


## Predicting functional motions of a single structure

```{r}
adk <- read.pdb("6s36")
adk
```

```{r}
# Perform flexiblity prediction
m <- nma(adk)
```

```{r}
plot(m)
```


```{r}
mktrj(m, file="adk_m7.pdb")
```


I ran these cmds in the R brain/console:

# Install packages in the R console NOT your Rmd/Quarto file

install.packages("bio3d")
install.packages("devtools")
install.packages("BiocManager")

BiocManager::install("msa")
devtools::install_bitbucket("Grantlab/bio3d-view")



>Q10. Which of the packages above is found only on BioConductor and not CRAN?

- msa

>Q11. Which of the above packages is not found on BioConductor or CRAN?:

- bio3d-view

>Q12. True or False? Functions from the devtools package can be used to install packages from GitHub and BitBucket?

- True

```{r}
library(bio3d)
aa <- get.seq("1ake_A")
aa
```
Here we will find and analyze all ADK structures in the PDB database

We will start with a single database accession id: "1ake_A"

>Q13. How many amino acids are in this sequence, i.e. how long is this sequence? 

```{r}
ncol(aa$ali)
```

- 214 amino acids


```{r}
# Blast or hmmer search 
#b <- blast.pdb(aa)
```

```{r}
# Plot a summary of search results
#hits <- plot(b)
```

```{r}
# List out some 'top hits'
#head(hits$pdb.id)
```

```{r}
pdb.id <- c('1AKE_A','6S36_A','6RZE_A','3HPR_A','1E4V_A','5EJE_A','1E4Y_A','3X2S_A','6HAP_A','6HAM_A','4K46_A','3GMT_A','4PZL_A')
```


```{r}
# Download releated PDB files
files <- get.pdb(pdb.id, path="pdbs", split=TRUE, gzip=TRUE)
```

Next we will use the pdbaln() function to align and also optionally fit (i.e. superpose) the identified PDB structures

```{r}
# Align releated PDBs
pdbs <- pdbaln(files, fit = TRUE, exefile="msa")
```
```{r}
pdbs
```



```{r}
# Vector containing PDB codes for figure axis
ids <- basename.pdb(pdbs$id)

# Draw schematic alignment
plot(pdbs, labels=ids)
```

```{r}
anno <- pdb.annotate(ids)
unique(anno$source)
```

```{r}
anno
```

```{r}
# Perform PCA
pc.xray <- pca(pdbs)
plot(pc.xray)
```

```{r}
# Calculate RMSD
rd <- rmsd(pdbs)

# Structure-based clustering
hc.rd <- hclust(dist(rd))
grps.rd <- cutree(hc.rd, k=3)

plot(pc.xray, 1:2, col="grey50", bg=grps.rd, pch=21, cex=1)
```


```{r}
uniprot <- 248838887
pdb <-        195610

pdb/uniprot * 100
```


```{r}
# Visualize first principal component
pc1 <- mktrj(pc.xray, pc=1, file="pc_1.pdb")
```

```{r}
#Plotting results with ggplot2
library(ggplot2)
library(ggrepel)

df <- data.frame(PC1=pc.xray$z[,1], 
                 PC2=pc.xray$z[,2], 
                 col=as.factor(grps.rd),
                 ids=ids)

p <- ggplot(df) + 
  aes(PC1, PC2, col=col, label=ids) +
  geom_point(size=2) +
  geom_text_repel(max.overlaps = 20) +
  theme(legend.position = "none")
p
```

```{r}
# NMA of all structures
modes <- nma(pdbs)
```

```{r}
plot(modes, pdbs, col=grps.rd)
```


>Q14. What do you note about this plot? Are the black and colored lines similar or different? Where do you think they differ most and why?

- The black and colored lines are different, the black lines have relatively low peaks, whereas the colored lines have large peaks throughout the plot. They differ most between residue numbers ~40-60, and ~120-154. 












